### DISCUSSION TITLE
[答疑] 为什么不推荐cq码和http接口用于实际开发呢

### DISCUSSION THREAD

**Discussion by MliKiowa** [NapNeko/NapCatQQ#518] [General]
事实上cq码本身是一种荒谬的想法 但在那个时候具有一定合理性

其荒谬性在于何处 cq码不是一种通用的数据格式 仅仅在ob协议中流行 因此各语言也并没有非常标准的解析库与序列化库 其不同数据混合就类似于sql码拼接这种效果 大多数人习惯以文本的方式处理 导致非常容易引发bug 异常与漏洞 对于复杂结构体 也无法直观展示

为什么http也不推荐呢 http具有简单方便的使用 在部分情景非常有效 但是在大多数开发中 插件是功能复杂的 多元的 

而http呈现的是单工模式 如果需要监听event还需要httppost 这样的模式实现双工其效率与配置复杂性可想而知 ws可以进行异步与双工的特性 让复杂开发中更加高效 稳定

但是总结起来http/ws并不是非常好的东西 onebot也不是合格的标准 如今缺点若能忍受 也就还好

**Comment by 837951602** [comment]
ws或者双http都支持双方互call，还行吧
通讯每次都要序列号/反序列化倒是个 性能问题，不过好像也没什么解决方案
看起来复杂的操作已经不太支持cq码了，到处是[object Object]
